### ISSUE TITLE
Using DISK for aerial images

### ISSUE THREAD

**Issue by tudipffmgt** [cvlab-epfl/disk#19]
Thank you for your amazing work.
I am using DISK quite successfully on aerial images, however, these mostly have high resolutions (talking about 16000 x 16000 pixels).
I am using the original images as .tif files in your proposed inference workflow.

If I use python `detect.py --height 3200 --width 3200` (maximum for our GPUs) is a high accuracy of the final keypoints in the original image with full resolution still given?
Or would you propose a different workflow?

**Comment by jatentaki** [comment]
TL;DR: I don't really know the answer but if you see it working reasonably, I don't think it's missing so much of its potential. Below I try to list out the possible issues with DISK in your application so you know where to search for improvements.

First, I'm very happy to see DISK being this robust to distribution shift - it really never was trained on aerial imagery. In general the network is a convnet so it doesn't make a difference how big the input image is. I have used DISK on 2048x2048 images (2.5x larger than training) and it seemed to work just fine, I didn't benchmark it rigorously though. There are two problems that could arise from large resolutions:
1. If you do not cap the number of detections or keep it very high, you may eventually "saturate" the space of descriptors - at some point you inevitably will have spurious nearest neighbors in matching because there are only so many pairs of points you can reasonably embed in the 128-dimensional space of descriptors. That said, I believe this ceiling is high, having made some experiments showing that reducing the dimensionality to 64 makes almost no difference with 8k detections. Finally, if you do keep detection threshold very high, you're likely to run into memory issues during matching.
2. If you do choose to cap the number of detections, the question is how well (and reliably) are they ranked. With 3200x3200 and a grid size of 8x8, you can theoretically get up to 160k detections in an image. Let's say you cap that to 8k, that's 1 in 20 detections kept. As you can imagine, the chances that some keypoint was detected in both images but then discarded in one image and not the other can be quite high. It really depends on how well "calibrated" the detection scores are, and the fact they are at all calibrated is already a nice empirical surprise - our training procedure doesn't directly promote that.

That said, my bigger concerns with using DISK for aerial photography would be 
1. Distribution shift. Obvious - the network has never seen any aerial photography. There could be a lot to gain by fine-tuning it on your domain data.
2. Rotation invariance. Unless you always have the images oriented correctly, the descriptors are quite big on using "up-is-up" as a cue (see the figure below).
<img width="866" alt="image" src="https://user-images.githubusercontent.com/22394841/216764091-cdd67ef8-2447-4f8b-8a36-ba5c2819fe9a.png">

**Comment by tudipffmgt** [comment]
Dear Michal,

thank you for your fast and elaborate reply.
Really interesting for us as we experienced the issues you pointed out.
Rotational invariance is tackled by us using variations of the original images turned by 90°, 180° and 270°.
We were already quite surprised that the detection works so well on aerial images, though trained on the MegaDepth dataset.
Additionally, I did a lot of parameter testing and tested with different image sizes and number of detected features.
I experienced that when I set the number of features to unlimited I got worse results and in our tests for the large aerial images the best setup was, `--height 3200 --width 3200 --n 32768 nms`
If you are interested I will send you the final paper when it is published.

Thanks and best regards!
